File: app/src/main/java/jp/hazuki/yuzubrowser/toolbar/main/ProgressToolBar.java
package jp.hazuki.yuzubrowser.toolbar.main;

import android.content.Context;
import android.graphics.PorterDuff;
import android.view.View;
import android.widget.ProgressBar;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.ActionCallback;
import jp.hazuki.yuzubrowser.settings.data.AppData;
import jp.hazuki.yuzubrowser.settings.data.ThemeData;
import jp.hazuki.yuzubrowser.tab.manager.MainTabData;
import jp.hazuki.yuzubrowser.toolbar.ButtonToolbarController;
import jp.hazuki.yuzubrowser.toolbar.ToolbarManager.RequestCallback;
import jp.hazuki.yuzubrowser.utils.DisplayUtils;

public class ProgressToolBar extends ToolbarBase {
    private final ProgressBar progressBar;

    public ProgressToolBar(Context context, ActionCallback action_callback, RequestCallback request_callback) {
        super(context, AppData.toolbar_progress, R.layout.toolbar_progress, request_callback);

        int TOOLBAR_SIZE_Y = DisplayUtils.convertDpToPx(context, AppData.toolbar_progress.size.get());

        progressBar = (ProgressBar) findViewById(R.id.progressBar);
        ButtonToolbarController.settingButtonSizeHeight(progressBar, TOOLBAR_SIZE_Y);
    }

    @Override
    public void onPreferenceReset() {
        super.onPreferenceReset();
    }

    @Override
    public void applyTheme(ThemeData themedata) {
        super.applyTheme(themedata);
        if (themedata != null && themedata.progressIndeterminateColor != 0)
            progressBar.getIndeterminateDrawable().setColorFilter(themedata.progressIndeterminateColor, PorterDuff.Mode.SRC_ATOP);
        else
            progressBar.getIndeterminateDrawable().setColorFilter(null);

        if (themedata != null && themedata.progressColor != 0)
            progressBar.getProgressDrawable().setColorFilter(themedata.progressColor, PorterDuff.Mode.SRC_ATOP);
        else
            progressBar.getProgressDrawable().setColorFilter(null);
    }

    public void notifyChangeWebState(MainTabData data) {
        super.notifyChangeWebState(data);
        if (data != null)
            changeProgress(data);
    }

    public void changeProgress(MainTabData data) {
        int p = data.mProgress;
        boolean inload = data.isInPageLoad();

        if (p == 100 || !inload) {
            progressBar.setVisibility(View.INVISIBLE);
            if (progressBar.isIndeterminate()) progressBar.setIndeterminate(false);//need this?
        } else if (p <= 0) {
            progressBar.setVisibility(View.VISIBLE);
            if (!progressBar.isIndeterminate()) progressBar.setIndeterminate(true);
        } else {
            progressBar.setVisibility(View.VISIBLE);
            if (progressBar.isIndeterminate()) progressBar.setIndeterminate(false);
            progressBar.setProgress(p);
        }
    }
}

File: legacy/src/main/java/jp/hazuki/yuzubrowser/legacy/toolbar/main/ProgressToolBar.kt
/*
 * Copyright (C) 2017-2019 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.legacy.toolbar.main

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import androidx.core.graphics.BlendModeColorFilterCompat.createBlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import jp.hazuki.yuzubrowser.core.utility.extensions.convertDpToPx
import jp.hazuki.yuzubrowser.legacy.databinding.ToolbarProgressBinding
import jp.hazuki.yuzubrowser.legacy.tab.manager.MainTabData
import jp.hazuki.yuzubrowser.legacy.toolbar.ButtonToolbarController
import jp.hazuki.yuzubrowser.ui.settings.AppPrefs
import jp.hazuki.yuzubrowser.ui.theme.ThemeData

class ProgressToolBar(context: Context, request_callback: RequestCallback) : ToolbarBase(context, AppPrefs.toolbar_progress, request_callback) {
    private val binding = ToolbarProgressBinding.inflate(LayoutInflater.from(context), this, true)

    init {
        val toolbarSizeY = context.convertDpToPx(AppPrefs.toolbar_progress.size.get())

        ButtonToolbarController.settingButtonSizeHeight(binding.progressBar, toolbarSizeY)
    }

    override fun applyTheme(themeData: ThemeData?) {
        super.applyTheme(themeData)

        if (themeData != null && themeData.progressIndeterminateColor != 0) {
            binding.progressBar.indeterminateDrawable.colorFilter =
                createBlendModeColorFilterCompat(themeData.progressIndeterminateColor, BlendModeCompat.SRC_ATOP)
        } else {
            binding.progressBar.indeterminateDrawable.colorFilter = null
        }

        if (themeData != null && themeData.progressColor != 0) {
            binding.progressBar.progressDrawable.colorFilter =
                createBlendModeColorFilterCompat(themeData.progressColor, BlendModeCompat.SRC_ATOP)
        } else {
            binding.progressBar.progressDrawable.colorFilter = null
        }
    }

    override fun notifyChangeWebState(data: MainTabData?) {
        super.notifyChangeWebState(data)
        if (data != null)
            changeProgress(data)
    }

    fun changeProgress(data: MainTabData) {
        val p = data.mProgress

        val progressBar = binding.progressBar
        if (p == 100 || !data.isInPageLoad) {
            progressBar.visibility = View.INVISIBLE
            if (progressBar.isIndeterminate) progressBar.isIndeterminate = false//need this?
        } else if (p <= 0) {
            progressBar.visibility = View.VISIBLE
            if (!progressBar.isIndeterminate) progressBar.isIndeterminate = true
        } else {
            progressBar.visibility = View.VISIBLE
            if (progressBar.isIndeterminate) progressBar.isIndeterminate = false
            progressBar.progress = p
        }
    }
}
